# Exam - bioinformatics for population genomics
## 2024-09-04

Please read carefully all instructions ! 

You must have created a GitHub account and invited `ericpante` to it. 
Please prepare a main file in MarkDown to explain your answers and provide your code. 
Please add your scripts as separate files in the GitHub repository. 
All code must be able to run on Ubuntu linux, as set up in the informatics labs at IUEM. 
The raw data consist in the following files: `SRR1754731, SRR1754726, SRR1754725, SRR1754723, SRR1754724, SRR1754722, SRR1754721`, 
already stored in `/shared/projects/tp_tps_iuem_bioinfo_166600/`.


## Useful links
- NCBI tools and databases are [here](https://blast.ncbi.nlm.nih.gov/Blast.cgi?CMD=Web&PAGE_TYPE=BlastDocs&DOC_TYPE=Download)
- `trimmomatic` is [here](http://www.usadellab.org/cms/?page=trimmomatic)
- `FastQC` is [here](http://www.usadellab.org/cms/?page=trimmomatic)
- `ipyrad` manual is [here](https://ipyrad.readthedocs.io/en/master/ )
- the `markdown` guide is [here](https://www.markdownguide.org)
- `multiqc` is [here](https://www.google.com/search?client=safari&rls=en&q=multiqc&ie=UTF-8&oe=UTF-8)
- GitHub is this site :-)

## Useful tips

You may want to work from the informatics lab or from a live ubuntu USB session, but make sure your code runs in the informatics lab. 
Get a `bash` shell cheat sheet to help you. Set up your working directory as such: `~/bioinfo` is your main directory. 
I recommand that you use the IFB cluster to take advantage of preinstalled `modules` (you know how to search for them).

## Exercise 1. 

Prepare your `fastq` files (see above). Retrieve and set-up the ncbi ITS database for Eukaryotes. 
Use the IFB cluster to run `fastqc`on all files. 
Run `trimmomatic` and clean the reads with the settings provided in the exemple command line for SE reads.
Re-run fastQC on the same files. 
Then run `multiqc` and prepare a summary table in `markdown` with the result of `fastqc` before and after cleaning by `trimmomatic`. 
Please explain the percentage of data removed and explain. 
Then, for one of the `fastq` files (chose one), search reads matching ITS sequences using `blast`. 

## Exercice 2. 

Use the cleaned data from exercise 1 to run `ipyrad`. Install `ipyrad` and dependencies. Create a `params` file for SE sdRAD reads obtained using Pst1. Use this file as a template: https://github.com/ericpante/bioinfo/blob/main/param-test.txt. Construct a catalog of loci with the default settings, branch your analysis (see `ipyrad` manual) and rerun only changing the `clust_threshold` parameter to 90%, 95%, 99%. Summarize as a `markdown`table describe the results; propose an interpretation for the results.
